src/api/kamino/transactions.ts

```
@@ -68,55 +69,119 @@ async function fetchTransactions(
+    const strategy = await redis.cacheFetchJson(
+      redisKey,
+      () => kamino.getStrategyByAddress(new PublicKey(strategyPubkey.pubKey)),
+      {
+        cacheExpiryType: CacheExpiryType.ExpireInSeconds,
+        cacheExpirySeconds: STRATEGIES_EXPIRY_IN_SECONDS,
+      }
+    );
+    if (!strategy) {
+      logger.error(`Could not find strategy ${strategyPubkey.pubKey}`);
       continue;
     }
-    const shares = instruction.txBalances[userSharesAta.pubKey].token;
-    const tokenA = instruction.txBalances[tokenAAta.pubKey].token;
-    const tokenB = instruction.txBalances[tokenBAta.pubKey].token;
 
-    const { numberOfShares, tokenAAmount, tokenBAmount } = getTokenAmounts(shares, tokenA, tokenB);
+    const userSharesAta = await getAssociatedTokenAddress(new PublicKey(strategy.sharesMint), shareholder);
+    const tokenAAta = await getAssociatedTokenAddress(new PublicKey(strategy.tokenAMint), shareholder);
+    const tokenBAta = await getAssociatedTokenAddress(new PublicKey(strategy.tokenBMint), shareholder);
+
+    const shares = instruction.txBalances[userSharesAta.toString()].token;
+    const tokenA = instruction.txBalances[tokenAAta.toString()].token;
+    const tokenB = instruction.txBalances[tokenBAta.toString()].token;
 
     const { tokenAName, tokenBName, tokenAPrice, tokenBPrice } = await getTokenInfo(
       timestamp,
-      tokenA,
       config,
-      tokenB,
-      env
+      env,
+      new PublicKey(strategy.tokenAMint),
+      new PublicKey(strategy.tokenBMint)
     );
 
+    const tokenAIsSol = strategy.tokenAMint.toString() === solMint.toString();
+    const tokenBIsSol = strategy.tokenBMint.toString() === solMint.toString();
+
+    const numberOfShares = getTokenDifferenceFromPrePostBalance(shares);
+    if (numberOfShares === null) {
+      logger.error(`Could not get token difference from token pre/post balances for number of shares`);
+      continue;
```

throw or return error, no silent errors here please